### ISSUE TITLE
How can I make my Zig wasm4 lib testable?

### ISSUE THREAD

**Issue by freiguy1** [aduros/wasm4#225]
I'm admittedly new to the zig world, but I thought this would be a great way to learn. Since I'm new, it's nice to be able to run small isolated bits of code and check for correctness: a unit test. But running `zig build test` doesn't work and `zig test src/main.zig` gets a bit farther, but I get an error related to wasm4 interop:
```
[me@arch-vbox w4-hello-world]$ zig test src/main.zig
ld.lld: error: undefined symbol: rect
>>> referenced by main.zig:65 (/home/me/dev/zig/w4-hello-world/src/main.zig:65)
>>>               src/zig-cache/o/2284a29267f88658f44b6e110941a408/test.o:(update)
>>> referenced by main.zig:70 (/home/me/dev/zig/w4-hello-world/src/main.zig:70)
>>>               src/zig-cache/o/2284a29267f88658f44b6e110941a408/test.o:(update)
>>> referenced by main.zig:75 (/home/me/dev/zig/w4-hello-world/src/main.zig:75)
>>>               src/zig-cache/o/2284a29267f88658f44b6e110941a408/test.o:(update)
>>> referenced 1 more times
```

While this is probably a zig question, it could also get included in the zig template so future devs could test!

**Comment by desttinghim** [comment]
There would have to be a stub implementation of wasm4 that provides the functions. It should be pretty easy to write, I may even get around to it myself.

**Comment by freiguy1** [comment]
That makes sense! I could attempt something, but I guess I'm not sure where to start. Would the stub wasm4 be written in c, or could I do it in zig as well? I wonder if there's a similar setup for some other project out there I could mimic. Maybe another route I could  go is to have a different lib/package (whatever the proper term is) which contains my logic (and no wasm4 interop) where I could house some tests. Then my game's main.zig would depend on that.

**Comment by freiguy1** [comment]
Ooh, maybe I can use this from the zig docs:
>Use the compile variable `@import("builtin").is_test` to detect a test build

**Comment by joshgoebel** [comment]
I think all you need to do is stub out the wasm4 stuff (not sure why Zig wouldn't work)... (assuming you aren't trying to test that)... all those functions just need to link and compiler, but they don't need to do anything.

**Comment by freiguy1** [comment]
Okay I'll use zig for the stub. I think it's the _linking_ part that I'm confused on. I guess I'm not sure how to link to the stub when testing (I'm not even sure if my question has the correct terminology, since I'm a bit naive to how linking works).

My first attempt is to import wasm4.zig this way:
`const w4 = if (@import("builtin").is_test) @import("wasm4stub.zig") else @import("wasm4.zig");`

And I have gotten tests to work correctly this way! However it feels a little weird to tamper with src/main.zig in order to test correctly. It seems like it'd be cleaner in build.zig.

**Comment by desttinghim** [comment]
@freiguy1 you can specify a main file for testing by using this in your build.zig:

```zig
b.addTest("test.zig")
```
Then you can pull in the files you need like normal. If you have a file where you want the tests to run but you don't need to use it in test.zig, use the following:

```zig
comptime {
    _ = @import("anything.zig");
}
```

**Comment by freiguy1** [comment]
Okay then this new test.zig file would no dependency on wasm4 (aka wouldn't `@import` it) and no dependency on main.zig (which depends on wasm4), right?

**Comment by desttinghim** [comment]
Yeah, pretty much. As long as all your wasm4 reliant code is in main this will work. Now that I think of it though, you may just want to make a w4.zig that has `usingnamespace if (@import("builtin").is_test) @import("wasm4stub.zig") else @import("wasm4.zig");` and import that wherever you want to use wasm4. I haven't tested this code but I think it should work.

**Comment by freiguy1** [comment]
Ooh using `usingnamespace`, I think I got something that's pretty optimal (not much repeated code and unobtrusive in main.zig) https://gist.github.com/freiguy1/aec1cbd670b3be6d52b60e22ff69614b

It's all in one file.
